README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mpsTemplates

<!-- badges: start -->

[![license](https://img.shields.io/badge/license-MIT%20+%20file%20LICENSE-lightgrey.svg)](https://choosealicense.com/)
[![R-CMD-check](https://github.com/mps9506/mps-templates/workflows/R-CMD-check/badge.svg)](https://github.com/mps9506/mps-templates/actions)

<!-- badges: end -->

``` r
library(ggplot2)
library(mpsTemplates)
ggplot(cars) +
  geom_point(aes(speed, dist)) +
  theme_mps_noto() +
  labs(title = "Cars", subtitle = "Speed and stopping distance",
       caption = "Ezekiel, M (1930) Methods of Correlation Analysis",
       x = "Speed [mph]",
       y = "Stopping Distance [ft]")
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

## pkgdown

The package includes some pkgdown template files intended for use by my
packages.

To use, add/edit \_pkgdown.yaml in the package root directory:

``` yaml
template:
  package: mpsTemplates
  bootstrap: 5

navbar:
  type: light
  bg: light
  structure:
    right: [search, github]
```

AND add the following to the DESCRIPTION:

    Config/Needs/website: mps9506/mpsTemplates

Use rlibs Github Action to build the pkgdown documentation:

``` r
usethis::use_github_action("pkgdown")
```

## rmarkdown

The package includes the following rmarkdown templates:

-   *mps pdf Cormorant template*: Intended for pdf reports.

These are available in the RStudio New -> Rmarkdown selector after
installing this package.

## bookdown

The package include the following project:

-   *mps Bookdown Single Page Website*: Intended for single page
    websites hosted on github pages

These are available by loading the package and selecting “New Project”
-> “New Directory” -> “mps Github Pages Website”


NEWS.md
# mpsTemplates 0.2.0

* darker background on .pre code blocks and appropriate code highlighting

# mpsTemplates 0.1.0

* Initial release


R/init_fonts.R
init_fonts <- function() {
  ## register fonts
  NotoSansDisplay_Condensed_Regular <- fs::path_package("mpsTemplates", "fonts", "NSD_C", "NotoSansDisplay_Condensed-Regular.ttf")
  NotoSansDisplay_Condensed_Bold <- fs::path_package("mpsTemplates", "fonts", "NSD_C", "NotoSansDisplay_Condensed-Bold.ttf")
  NotoSansDisplay_Condensed_Italic <- fs::path_package("mpsTemplates", "fonts", "NSD_C", "NotoSansDisplay_Condensed-Italic.ttf")
  NotoSansDisplay_Condensed_BoldItalic <- fs::path_package("mpsTemplates", "fonts", "NSD_C", "NotoSansDisplay_Condensed-BoldItalic.ttf")

  systemfonts::register_font(
    "NotoSansDisplay_Condensed",
    plain = NotoSansDisplay_Condensed_Regular,
    bold = NotoSansDisplay_Condensed_Bold,
    italic = NotoSansDisplay_Condensed_Italic,
    bolditalic = NotoSansDisplay_Condensed_BoldItalic
  )

  NotoSansDisplay_Regular <- fs::path_package("mpsTemplates", "fonts", "NSD", "NotoSansDisplay-Regular.ttf")
  NotoSansDisplay_Bold <- fs::path_package("mpsTemplates", "fonts", "NSD", "NotoSansDisplay-Bold.ttf")
  NotoSansDisplay_Italic <- fs::path_package("mpsTemplates", "fonts", "NSD", "NotoSansDisplay-Italic.ttf")
  NotoSansDisplay_BoldItalic <- fs::path_package("mpsTemplates", "fonts", "NSD", "NotoSansDisplay-BoldItalic.ttf")

  systemfonts::register_font(
    "NotoSansDisplay",
    plain = NotoSansDisplay_Regular,
    bold = NotoSansDisplay_Bold,
    italic = NotoSansDisplay_Italic,
    bolditalic = NotoSansDisplay_BoldItalic
  )

  NotoSansDisplay_SC_Medium <- fs::path_package("mpsTemplates", "fonts", "NSD_SC", "NotoSansDisplay_SemiCondensed-Medium.ttf")
  NotoSansDisplay_SC_Bold <- fs::path_package("mpsTemplates", "fonts", "NSD_SC", "NotoSansDisplay_SemiCondensed-Bold.ttf")
  NotoSansDisplay_SC_Italic <- fs::path_package("mpsTemplates", "fonts", "NSD_SC", "NotoSansDisplay_SemiCondensed-MediumItalic.ttf")
  NotoSansDisplay_SC_BoldItalic <- fs::path_package("mpsTemplates", "fonts", "NSD_SC", "NotoSansDisplay_SemiCondensed-BoldItalic.ttf")

   systemfonts::register_font(
    "NotoSansDisplay_SC",
    plain = NotoSansDisplay_SC_Medium,
    bold = NotoSansDisplay_SC_Bold,
    italic = NotoSansDisplay_SC_Italic,
    bolditalic = NotoSansDisplay_SC_BoldItalic
  )

  }


R/theme_mps_noto.R
#' an opinionated [ggplot2] theme
#'
#' Provides a [ggplot2] theme.
#'
#' @param base_family defaults to \code{"NotoSansDisplay_Condensed"}. Which is
#'   provided in the package. Generally, use a sensible sans serif font that is
#'   available on your system.
#' @param base_size default text size in px
#' @param base_line_size default line size
#' @param base_rect_size default rect size
#' @import ggplot2
#' @export
#' @details Uses Google's Noto Sans Display font by default. Using the
#'   [ragg::agg_png()] or related graphics device should ensure the font is
#'   properly detected and used.
#' @examples
#' library(ggplot2)
#' library(ragg)
#'
#' file <- tempfile(fileext = '.png')
#' agg_png(file)
#' ggplot(cars) +
#'   geom_point(aes(speed, dist)) +
#'   theme_mps_noto() +
#'   labs(title = "Cars", subtitle = "Speed and stopping distance",
#'        caption = "Ezekiel, M (1930) Methods of Correlation Analysis",
#'        x = "Speed [mph]",
#'        y = "Stopping Distance [ft]")
#' dev.off()
theme_mps_noto <- function(base_size = 11,
                             base_family = "NotoSansDisplay_Condensed",
                             base_line_size = 0.5,
                             base_rect_size = 0.5) {
  half_line <- base_size / 2L

  ggplot2::theme(
    # main attributes

    line = ggplot2::element_line(colour = "#000000",
                                 size = base_line_size,
                                 linetype = 1L,
                                 lineend = "butt"),
    rect = ggplot2::element_rect(fill = "#FFFFFF",
                                 colour = "#000000",
                                 size = base_rect_size,
                                 linetype = 1L),
    text = ggplot2::element_text(family = base_family,
                                 face = "plain",
                                 colour = "#000000",
                                 size = base_size,
                                 hjust = 0.5,
                                 vjust = 0.5,
                                 angle = 0,
                                 lineheight = 0.9,
                                 margin = ggplot2::margin(),
                                 debug = FALSE),

    # Plot Attributes

    plot.tag = ggplot2::element_text(size = base_size * 1.5,
                                     hjust = 0L,
                                     vjust = 0L,
                                     face = "bold",
                                     margin = ggplot2::margin(b = 10L)),
    plot.tag.position = "topleft",
    plot.title = ggplot2::element_text(size = base_size * 12 / 8.5,
                                       hjust = 0L,
                                       vjust = 0L,
                                       family = base_family,
                                       face = "bold",
                                       margin = ggplot2::margin(b = 10L)),
    plot.title.position = "plot",
    plot.subtitle = ggplot2::element_text(size = base_size * 9.5 / 8.5,
                                          hjust = 0L,
                                          vjust = 0L,
                                          family = base_family,
                                          margin = ggplot2::margin(b = 10L)),
    plot.caption = ggplot2::element_text(size = base_size * 7 / 8.5,
                                         hjust = 1L,
                                         vjust = 1L,
                                         margin = ggplot2::margin(t = half_line * 0.9)),
    plot.caption.position = "plot",
    plot.background = NULL,

    plot.margin = ggplot2::margin(t = half_line,
                                  r = base_line_size * 24,
                                  b = half_line,
                                  l = half_line),

    # axis attributes

    axis.text = ggplot2::element_text(size = base_size),
    axis.text.x = ggplot2::element_text(vjust = 1, margin = ggplot2::margin(t = 4L)),
    axis.text.y = ggplot2::element_text(),
    axis.text.x.top = NULL,
    axis.text.y.right = NULL,
    axis.ticks.length.x = NULL,
    axis.ticks.length.x.top = NULL,
    axis.ticks.length.x.bottom = NULL,
    axis.ticks.length.y = NULL,
    axis.ticks.length.y.left = NULL,
    axis.ticks.length.y.right = NULL,

    axis.title = ggplot2::element_text(size = base_size),
    axis.title.x = ggplot2::element_text(margin = ggplot2::margin(t = 8L)),
    axis.title.y = ggplot2::element_text(angle = 90L,
                                         margin = ggplot2::margin(r = 4L)),
    axis.title.x.top = NULL,
    axis.title.y.right = NULL,

    axis.ticks = ggplot2::element_line(),
    axis.ticks.length = ggplot2::unit(4L, "pt"),
    axis.ticks.x = ggplot2::element_line(colour = NULL,
                                         size = NULL,
                                         linetype = NULL,
                                         lineend = NULL),
    axis.ticks.y = ggplot2::element_line(colour = NULL,
                                         size = NULL,
                                         linetype = NULL,
                                         lineend = NULL),

    axis.line = ggplot2::element_line(),
    axis.line.x = ggplot2::element_line(),
    axis.line.x.bottom = element_blank(),
    axis.line.x.top = element_blank(),
    axis.line.y = ggplot2::element_line(),
    axis.line.y.left = element_blank(),
    axis.line.y.right = element_blank(),

    # legend attributes

    legend.background = ggplot2::element_blank(),

    legend.spacing = ggplot2::unit(20L, "pt"),
    legend.spacing.x = ggplot2::unit(4L, "pt"),
    legend.spacing.y = NULL,

    legend.key = ggplot2::element_blank(),
    legend.key.size = ggplot2::unit(10L, "pt"),
    legend.key.height = NULL,
    legend.key.width = NULL,

    legend.text = ggplot2::element_text(size = base_size * 9.5 / 8.5,
                                        vjust = 0.5),
    legend.text.align = NULL,
    legend.title = ggplot2::element_text(),
    legend.title.align = NULL,

    legend.position = "bottom",
    legend.direction = "horizontal",
    legend.justification = NULL,
    legend.margin = ggplot2::margin(t = 6L, r = 0L, b = 6L, l = 0L, "pt"),

    legend.box = "horizontal",
    legend.box.margin = NULL,
    legend.box.background = NULL,
    legend.box.spacing = NULL,

    # panel attributes

    panel.background = ggplot2::element_blank(),
    panel.border = ggplot2::element_rect(fill = NA),
    panel.ontop = FALSE,

    panel.spacing = ggplot2::unit(6L, "pt"),
    panel.spacing.x = NULL,
    panel.spacing.y = NULL,

    panel.grid = NULL,
    panel.grid.major = ggplot2::element_line(),
    panel.grid.major.x = ggplot2::element_blank(),
    panel.grid.major.y = ggplot2::element_line(linetype = "dotted", color = "#d9d9d9"),
    panel.grid.minor = ggplot2::element_line(),
    panel.grid.minor.x = ggplot2::element_blank(),
    panel.grid.minor.y = ggplot2::element_blank(),

    # strip attributes (Faceting)

    strip.background = ggplot2::element_rect(fill = "#dedddd",
                                             colour = NA,
                                             size = 10L),
    strip.text = ggplot2::element_text(face = "bold",
                                       size = base_size * 9.5 / 8.5,
                                       margin = ggplot2::margin(t = 0L, r = 0L, b = 0L, l = 0L)),

    strip.text.x = ggplot2::element_text(margin = ggplot2::margin(t = 4.5, b = 4.5)),
    strip.text.y = ggplot2::element_text(angle = -90L,
                                         margin = ggplot2::margin(l = 4.5, r = 4.5)),

    strip.placement = "inside",
    strip.placement.x =  NULL,
    strip.placement.y =  NULL,

    strip.switch.pad.grid = ggplot2::unit(0.1, "cm"),
    strip.switch.pad.wrap = ggplot2::unit(0.1, "cm"),

    # create a complete format
    complete = TRUE

  )
}


R/theme_mps_noto_dark.R
#' an opinionated [ggplot2] theme
#'
#' Provides a [ggplot2] theme.
#'
#' @param base_family defaults to \code{"NotoSansDisplay_Condensed"}. Which is
#'   provided in the package. Generally, use a sensible sans serif font that is
#'   available on your system.
#' @param base_size default text size in px
#' @param base_line_size default line size
#' @param base_rect_size default rect size
#' @import ggplot2
#' @export
#' @details Uses Google's Noto Sans Display font by default. Using the
#'   [ragg::agg_png()] or related graphics device should ensure the font is
#'   properly detected and used.
#' @examples
#' library(ggplot2)
#' library(ragg)
#'
#' file <- tempfile(fileext = '.png')
#' agg_png(file)
#' ggplot(cars) +
#'   geom_point(aes(speed, dist)) +
#'   theme_mps_noto_dark() +
#'   labs(title = "Cars", subtitle = "Speed and stopping distance",
#'        caption = "Ezekiel, M (1930) Methods of Correlation Analysis",
#'        x = "Speed [mph]",
#'        y = "Stopping Distance [ft]")
#' dev.off()
theme_mps_noto_dark <- function(base_size = 11,
                                base_family = "NotoSansDisplay_SC",
                                base_line_size = 0.5,
                                base_rect_size = 0.5) {
  ## some defaults
  half_line <- base_size / 2L
  text_color <- "white"
  rect_color <- "#839496"
  body_bg <- "#073642"
  geom_color <- "#b58900"

  # ## set geom defaults
  # ggplot2::update_geom_defaults("point", list(colour = geom_color))
  # ggplot2::update_geom_defaults("line", list(colour = geom_color))
  # ggplot2::update_geom_defaults("area", list(colour = geom_color, fill = geom_color))
  # ggplot2::update_geom_defaults("rect", list(colour = geom_color, fill = geom_color))
  # ggplot2::update_geom_defaults("density", list(colour = geom_color, fill = geom_color))
  # ggplot2::update_geom_defaults("bar", list(colour = geom_color, fill = geom_color))
  # ggplot2::update_geom_defaults("col", list(colour = geom_color, fill = geom_color))
  # ggplot2::update_geom_defaults("text", list(colour = alpha(text_color, 0.75)))

  ggplot2::theme(
    # main attributes

    line = ggplot2::element_line(colour = alpha(text_color, 0.55),
                                 size = base_line_size,
                                 linetype = 1L,
                                 lineend = "butt"),
    rect = ggplot2::element_rect(fill = rect_color,
                                 colour = rect_color,
                                 size = base_rect_size,
                                 linetype = 1L),
    text = ggplot2::element_text(family = base_family,
                                 face = "plain",
                                 colour = alpha(text_color, 0.75),
                                 size = base_size,
                                 hjust = 0.5,
                                 vjust = 0.5,
                                 angle = 0,
                                 lineheight = 0.9,
                                 margin = ggplot2::margin(),
                                 debug = FALSE),

    # Plot Attributes

    plot.tag = ggplot2::element_text(size = base_size * 1.5,
                                     hjust = 0L,
                                     vjust = 0L,
                                     face = "bold",
                                     margin = ggplot2::margin(b = 10L)),
    plot.tag.position = "topleft",
    plot.title = ggplot2::element_text(color = alpha(text_color, 0.95),
                                       size = base_size * 12 / 8.5,
                                       hjust = 0L,
                                       vjust = 0L,
                                       family = base_family,
                                       face = "bold",
                                       margin = ggplot2::margin(b = 10L)),
    plot.title.position = "plot",
    plot.subtitle = ggplot2::element_text(color = alpha(text_color, 0.75),
                                          size = base_size * 9.5 / 8.5,
                                          hjust = 0L,
                                          vjust = 0L,
                                          family = base_family,
                                          margin = ggplot2::margin(b = 10L)),
    plot.caption = ggplot2::element_text(color = alpha(text_color, 0.75),
                                         size = base_size * 7 / 8.5,
                                         hjust = 1L,
                                         vjust = 1L,
                                         margin = ggplot2::margin(t = half_line * 0.9)),
    plot.caption.position = "plot",
    plot.background = element_rect(color = NA,
                                   fill = body_bg),

    plot.margin = ggplot2::margin(t = half_line,
                                  r = base_line_size * 24,
                                  b = half_line,
                                  l = half_line),

    # axis attributes

    axis.text = ggplot2::element_text(size = base_size),
    axis.text.x = ggplot2::element_text(vjust = 1, margin = ggplot2::margin(t = 4L)),
    axis.text.y = ggplot2::element_text(),
    axis.text.x.top = NULL,
    axis.text.y.right = NULL,
    axis.ticks.length.x = NULL,
    axis.ticks.length.x.top = NULL,
    axis.ticks.length.x.bottom = NULL,
    axis.ticks.length.y = NULL,
    axis.ticks.length.y.left = NULL,
    axis.ticks.length.y.right = NULL,

    axis.title = ggplot2::element_text(size = base_size),
    axis.title.x = ggplot2::element_text(margin = ggplot2::margin(t = 8L)),
    axis.title.y = ggplot2::element_text(angle = 90L,
                                         margin = ggplot2::margin(r = 4L)),
    axis.title.x.top = NULL,
    axis.title.y.right = NULL,

    axis.ticks = ggplot2::element_line(colour = alpha(rect_color, 0.4)),
    axis.ticks.length = ggplot2::unit(4L, "pt"),
    axis.ticks.x = ggplot2::element_line(colour = NULL,
                                         size = NULL,
                                         linetype = NULL,
                                         lineend = NULL),
    axis.ticks.y = ggplot2::element_line(colour = NULL,
                                         size = NULL,
                                         linetype = NULL,
                                         lineend = NULL),

    axis.line = ggplot2::element_line(),
    axis.line.x = ggplot2::element_line(),
    axis.line.x.bottom = element_blank(),
    axis.line.x.top = element_blank(),
    axis.line.y = ggplot2::element_line(),
    axis.line.y.left = element_blank(),
    axis.line.y.right = element_blank(),

    # legend attributes

    legend.background = ggplot2::element_blank(),

    legend.spacing = ggplot2::unit(20L, "pt"),
    legend.spacing.x = ggplot2::unit(4L, "pt"),
    legend.spacing.y = NULL,

    legend.key = ggplot2::element_blank(),
    legend.key.size = ggplot2::unit(10L, "pt"),
    legend.key.height = NULL,
    legend.key.width = NULL,

    legend.text = ggplot2::element_text(size = base_size * 9.5 / 8.5,
                                        vjust = 0.5),
    legend.text.align = NULL,
    legend.title = ggplot2::element_text(),
    legend.title.align = NULL,

    legend.position = "bottom",
    legend.direction = "horizontal",
    legend.justification = NULL,
    legend.margin = ggplot2::margin(t = 6L, r = 0L, b = 6L, l = 0L, "pt"),

    legend.box = "horizontal",
    legend.box.margin = NULL,
    legend.box.background = NULL,
    legend.box.spacing = NULL,

    # panel attributes

    panel.background = ggplot2::element_rect(fill = alpha("#1C2023", 0.75), colour = NA),
    panel.border = ggplot2::element_rect(color = alpha(rect_color, 0.5),
                                         fill = NA),
    panel.ontop = FALSE,

    panel.spacing = ggplot2::unit(6L, "pt"),
    panel.spacing.x = NULL,
    panel.spacing.y = NULL,

    panel.grid = NULL,
    panel.grid.major = ggplot2::element_line(),
    panel.grid.major.x = ggplot2::element_blank(),
    panel.grid.major.y = ggplot2::element_line(linetype = "dotted",
                                               color = alpha(rect_color, 0.5)),
    panel.grid.minor = ggplot2::element_line(),
    panel.grid.minor.x = ggplot2::element_blank(),
    panel.grid.minor.y = ggplot2::element_blank(),

    # strip attributes (Faceting)

    strip.background = ggplot2::element_rect(fill = "#dedddd",
                                             colour = NA,
                                             size = 10L),
    strip.text = ggplot2::element_text(face = "bold",
                                       size = base_size * 9.5 / 8.5,
                                       margin = ggplot2::margin(t = 0L, r = 0L, b = 0L, l = 0L)),

    strip.text.x = ggplot2::element_text(margin = ggplot2::margin(t = 4.5, b = 4.5)),
    strip.text.y = ggplot2::element_text(angle = -90L,
                                         margin = ggplot2::margin(l = 4.5, r = 4.5)),

    strip.placement = "inside",
    strip.placement.x =  NULL,
    strip.placement.y =  NULL,

    strip.switch.pad.grid = ggplot2::unit(0.1, "cm"),
    strip.switch.pad.wrap = ggplot2::unit(0.1, "cm"),

    # create a complete format
    complete = TRUE

  )
}


#' Change geom defaults from black to custom lights
#'
#' @export
noto_dark_geom_defaults <- function() {

  geoms <- c("abline", "area", "bar", "boxplot", "col", "crossbar",
             "density", "dotplot", "errorbar", "errorbar",
             "hline", "line", "linerange",
             "map", "path", "point", "polygon", "rect", "ribbon", "rug", "segment",
             "step", "tile", "violin", "vline")

  for (g in geoms) {
    update_geom_defaults(g, list(colour = "#96896B", fill = "#96896B"))
  }

  geoms <- c("text", "label")

  for (g in geoms) {
    update_geom_defaults(g, list(colour = alpha("white", 0.85), fill = alpha("white", 0.20)))
  }

}
